src/proto/h2/mod.rs

```diff
@@ -377,9 +377,21 @@ where
 
     fn poll_shutdown(
         mut self: Pin<&mut Self>,
-        _cx: &mut Context<'_>,
+        cx: &mut Context<'_>,
     ) -> Poll<Result<(), io::Error>> {
-        Poll::Ready(self.send_stream.write(&[], true))
+        if self.send_stream.write(&[], true).is_ok() {
+            return Poll::Ready(Ok(()))
+        }
+
+        Poll::Ready(Err(h2_to_io_error(
+            match ready!(self.send_stream.poll_reset(cx)) {
+                Ok(Reason::NO_ERROR) | Ok(Reason::CANCEL) | Ok(Reason::STREAM_CLOSED) => {
+                    return Poll::Ready(Ok(()))
```

Comment:
If there was no error or the stream was cancelled or closed, we just return `Ok(())`, does that make sense to you @seanmonstar?

Reply:
[Originally](https://httpwg.org/specs/rfc7540.html#ErrorCodes), `STREAM_CLOSED` is meant to signal that the sender did something wrong, sending a frame after it said it wouldn't. So that seems like it should be an error.

I kind of think `CANCEL` should also be returned, at least so the sender knows. The user could chose to ignore those, right?

Reply:
Done